specs/crux/beacon-chain.md

```diff
@@ -0,0 +1,247 @@
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+```
+
+##### `process_delegation`
+
+```python
+def process_delegation(state: BeaconState, delegation: Delegation) -> None:
+    message = delegation.message
+    current_epoch = get_current_epoch(state)
+    assert  message.epoch <= current.epoch <= message.epoch + MAX_DELEGATION_INCLUSION_DELAY
+
+    delegating_index = message.delegating_index
+    min_delegating_balance = MAX_EFFECTIVE_BALANCE + MIN_DEPOSIT_AMOUNT + DELEGATION_TRANSACTION_COST
+    assert state.balances[delegating_index] >= min_delegating_balance
+    amount = state.balances[delegating_index] - MAX_EFFECTIVE_BALANCE - DELEGATION_TRANSACTION_COST
+
+    delegating_validator = state.validators[delegating_index]
+    assert is_active_validator(delegating_validator, current_epoch)
+    assert not delegating_validator.slashed 
+
+    assert delegating_validator.delegate == delegating_index
+
+    delegating_pubkey = message.delegating_pubkey
+    assert hash(delegating_pubkey)[1:] == delegating_validator.withdrawal_credentials[1:]
+
+    domain = compute_domain(DOMAIN_DELEGATION)
+    signing_root = compute_signing_root(message, domain)
+    assert bls.Verify(delegating_pubkey, signing_root, delegation.signature)
+
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    if pubkey not in validator_pubkeys:
+        index = len(state.validators)
+        state.validators.append(get_validator_from_delegation(message, amount))
+    else:
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        assert index > delegating_index
+        assert state.validators[index].exit_epoch == FAR_FUTURE_EPOCH
+
+    delegating_validator.delegate = index
+    proposer_index = get_beacon_proposer_index(state)
+    decrease_balance(state, delegating_validator, amount + DELEGATION_TRANSACTION_COST)
+    increase_balance(state, index, amount)
+    increase_balance(state, proposer_index, DELEGATION_TRANSACTION_COST)
+```
+
+### Epoch processing
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_delegation_transfers(state)  # [New in Crux]
+    process_registry_updates(state)   # [Modified in Crux]
+    process_slashings(state)
+    process_eth1_data_reset(state)
+    process_effective_balance_updates(state)
+    process_slashings_reset(state)
+    process_randao_mixes_reset(state)
+    process_historical_roots_update(state)
+    process_participation_flag_updates(state)
+    process_sync_committee_updates(state)
+```
+
+#### `process_delegation_transfers`
+
+```python
+def process_delegation_transfers(state: BeaconState) -> None:
+    delegating_validators = [(i, v) for i, v in enumerate(state.validators) if v.delegate > i and not v.slashed]
```

Comment:
Validators can't delegate to validators with index less than theirs?

Reply:
no, this is a cheap way to prevent loops.